File: webapp-conference-speakup/src/lib/websocket.ts
import { io, Socket } from "socket.io-client";
import { auth } from "@/lib/firebase";
import { WS_BASE_URL } from "@/config/constants";

class WebSocketService {
  private static instance: WebSocketService;
  private socket: Socket | null = null;
  private reconnectAttempts = 0;
  private readonly maxReconnectAttempts = 10;

  private constructor() {}

  static getInstance(): WebSocketService {
    if (!WebSocketService.instance) {
      WebSocketService.instance = new WebSocketService();
    }
    return WebSocketService.instance;
  }

  async connect(): Promise<Socket> {
    if (this.socket?.connected) return this.socket;

    const user = auth?.currentUser;
    const token = user ? await user.getIdToken() : undefined;

    this.socket = io(WS_BASE_URL, {
      auth: { token },
      transports: ["websocket", "polling"],
      reconnection: true,
      reconnectionAttempts: this.maxReconnectAttempts,
      reconnectionDelay: 1000,
      reconnectionDelayMax: 10000,
    });

    this.socket.on("connect", () => {
      this.reconnectAttempts = 0;
    });

    this.socket.on("disconnect", () => {
      this.reconnectAttempts++;
    });

    return this.socket;
  }

  disconnect() {
    this.socket?.disconnect();
    this.socket = null;
  }

  getSocket(): Socket | null {
    return this.socket;
  }

  // ─── Meeting events ───
  joinMeeting(meetingId: string) {
    this.socket?.emit("meeting:join", { meetingId });
  }
  leaveMeeting(meetingId: string) {
    this.socket?.emit("meeting:leave", { meetingId });
  }
  onParticipantJoined(cb: (data: unknown) => void) {
    this.socket?.on("meeting:participant-joined", cb);
  }
  onParticipantLeft(cb: (data: unknown) => void) {
    this.socket?.on("meeting:participant-left", cb);
  }
  onMeetingEnded(cb: () => void) {
    this.socket?.on("meeting:ended", cb);
  }

  // ─── Chat events ───
  joinChatRoom(chatRoomId: string) {
    this.socket?.emit("chat:join", { chatRoomId });
  }
  leaveChatRoom(chatRoomId: string) {
    this.socket?.emit("chat:leave", { chatRoomId });
  }
  sendMessage(chatRoomId: string, content: string, type = "text") {
    this.socket?.emit("chat:message", { chatRoomId, content, type });
  }
  onNewMessage(cb: (data: unknown) => void) {
    this.socket?.on("chat:new-message", cb);
  }
  onTyping(cb: (data: unknown) => void) {
    this.socket?.on("chat:typing", cb);
  }
  emitTyping(chatRoomId: string) {
    this.socket?.emit("chat:typing", { chatRoomId });
  }

  // ─── Notification events ───
  onNotification(cb: (data: unknown) => void) {
    this.socket?.on("notification:new", cb);
  }

  // ─── Generic listener ───
  on(event: string, cb: (data: unknown) => void) {
    this.socket?.on(event, cb);
  }
  off(event: string, cb?: (data: unknown) => void) {
    this.socket?.off(event, cb);
  }
}

export const ws = WebSocketService.getInstance();

File: Flutter-conference-speakup/lib/core/services/websocket.dart
import 'dart:async';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:socket_io_client/socket_io_client.dart' as io;
import 'package:flutter_conference_speakup/core/config/base_url.dart';
import 'package:logger/logger.dart';

final _log = Logger(printer: PrettyPrinter(methodCount: 0));

typedef SocketEventHandler = void Function(dynamic data);

class WebSocketService {
  static WebSocketService? _instance;
  io.Socket? _socket;
  final _eventControllers = <String, StreamController<dynamic>>{};
  bool _reconnecting = false;

  WebSocketService._();
  factory WebSocketService() => _instance ??= WebSocketService._();

  bool get isConnected => _socket?.connected ?? false;
  bool get isReconnecting => _reconnecting;

  /// Connect to WebSocket server with Firebase auth token.
  Future<void> connect() async {
    if (isConnected) return;

    String? token;
    final user = FirebaseAuth.instance.currentUser;
    if (user != null) {
      token = await user.getIdToken();
    }

    _socket = io.io(
      AppBaseUrl.wsUrl,
      io.OptionBuilder()
          .setTransports(['websocket'])
          .disableAutoConnect()
          .enableReconnection()
          .setReconnectionAttempts(10)
          .setReconnectionDelay(1000)
          .setReconnectionDelayMax(10000)
          .setAuth(token != null ? {'token': token} : {})
          .build(),
    );

    _socket!
      ..onConnect((_) {
        _log.i('[WS] Connected');
        _reconnecting = false;
        _emit('connection_status', true);
      })
      ..onDisconnect((_) {
        _log.w('[WS] Disconnected');
        _emit('connection_status', false);
      })
      ..onReconnect((_) {
        _log.i('[WS] Reconnected');
        _reconnecting = false;
        _emit('connection_status', true);
      })
      ..onReconnectAttempt((_) {
        _reconnecting = true;
      })
      ..onError((err) {
        _log.e('[WS] Error: $err');
        _emit('socket_error', err);
      })
      ..connect();
  }

  void disconnect() {
    _socket?.disconnect();
    _socket?.dispose();
    _socket = null;
    _reconnecting = false;
  }

  void emit(String event, [dynamic data]) {
    _socket?.emit(event, data);
  }

  void on(String event, SocketEventHandler handler) {
    _socket?.on(event, handler);
  }

  void off(String event) {
    _socket?.off(event);
  }

  /// Stream-based listener for reactive patterns.
  Stream<dynamic> stream(String event) {
    _eventControllers[event] ??= StreamController<dynamic>.broadcast();
    _socket?.on(event, (data) {
      _eventControllers[event]?.add(data);
    });
    return _eventControllers[event]!.stream;
  }

  void _emit(String event, dynamic data) {
    if (_eventControllers.containsKey(event)) {
      _eventControllers[event]?.add(data);
    }
  }

  /// Join a meeting room via WebSocket.
  void joinMeetingRoom(String meetingId) {
    emit('meeting:join', {'meetingId': meetingId});
  }

  /// Leave a meeting room via WebSocket.
  void leaveMeetingRoom(String meetingId) {
    emit('meeting:leave', {'meetingId': meetingId});
  }

  /// Join a chat room for real-time messages.
  void joinChatRoom(String chatRoomId) {
    emit('chat:join', {'chatRoomId': chatRoomId});
  }

  /// Leave a chat room.
  void leaveChatRoom(String chatRoomId) {
    emit('chat:leave', {'chatRoomId': chatRoomId});
  }

  void dispose() {
    disconnect();
    for (final ctrl in _eventControllers.values) {
      ctrl.close();
    }
    _eventControllers.clear();
    _instance = null;
  }
}
